### ISSUE TITLE
run more than one DDP model on one machine

### ISSUE THREAD

**Issue by FabianIsensee** [NVIDIA/apex#162]
Hi,
thank you so much for your input so far. I have now successfully converted my models to use DDP instead of DP and time per epoch has gone down from 1400s to 1140s when using 8 GPUs.
One problem that I was unable to solve so far is running more than one DDP model on the same machine:

```
Traceback (most recent call last):
  File "run/run_training_DPP.py", line 68, in <module>
    unpack_data=unpack, deterministic=deterministic, fp16=args.fp16)
  File "/home/fabian/PhD/meddec/meddec/model_training/distributed_training/nnUNetTrainerDPP.py", line 26, in __init__
    dist.init_process_group(backend='nccl', init_method='env://')
  File "/home/fabian/dl_venv_python3/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 354, in init_process_group
    store, rank, world_size = next(rendezvous(url))
  File "/home/fabian/dl_venv_python3/lib/python3.6/site-packages/torch/distributed/rendezvous.py", line 143, in _env_rendezvous_handler
    store = TCPStore(master_addr, master_port, world_size, start_daemon)
RuntimeError: Address already in use

```
(which is kind of to be expected given how DDP works). Do you know a workaround for this? We recently have gotten a dgx2 with 16 GPUs and I would like to run two different experiments in parallel, each using 8 GPUs.
Best,
Fabian

**Comment by mcarilli** [comment]
By different experiments do you mean you are manually launching two separate jobs with two separate invocations of 
`python -m multiproc --nproc_per_node=8 train.py ....`?
If so, I think you can make this work by additionally supplying unique MASTER_ADDR and MASTER_PORT environment variables to each job.  If you supply the appropriate args to `multiproc`, `multiproc` will set them up for you:
https://pytorch.org/docs/stable/distributed.html#launch-utility
```
python -m torch.distributed.launch --nproc_per_node=NUM_GPUS_YOU_HAVE
           --master_addr="192.168.1.1"
           --master_port=1234 YOUR_TRAINING_SCRIPT.py (--arg1 --arg2 --arg3
           and all other arguments of your training script)
```
^i took their multinode launch example and removed the multinode-specific args, since you're not doing a multinode launch, you're (I think) launching two jobs separately on a single node.

Another thing you will need to do is make sure `CUDA_VISIBLE_DEVICES` is different for each launch, so that the first launch uses gpus 0-7 and the second uses gpus 8-15.  What I would do is open two different terminals.  In the first terminal say
```
export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
python -m multiproc --nproc_per_node=8 --master_addr="IP for job 1" --master_port=<port for job 1> train.py ...
```
In the second terminal say
```
export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
python -m multiproc --nproc_per_node=8 --master_addr="IP for job 2" --master_port=<port for job 2> train.py ...
```

**Comment by mcarilli** [comment]
Let me know if the above approach works, I've never actually tried it on a DGX2, and it would be valuable information for other people.  This issue is not Pytorch-specific, but an issue with Pytorch distributed training in general.

**Comment by FabianIsensee** [comment]
Exactly, I want to run two different models on the same node, each using 8 GPUs. These models are completely independent trainings, they don't communicate with each other whatsoever.  There is no need to use all GPUs for one model, my GPUs are better utilized if I run several different models on the same machine. Previously I did that by running two models that used nn.DataParallel each (in combination with CUDA_VISIBLE_DEVICES as you outlined above).
How would I set the master_addr if both trainings run on the same machine? Then the IP would be identical wouldn't it? What ports are appropriate? Please excuse if these questions seem a bit dumb - I am completely new to this.
This problem is not limited to dgx2. My local workstation has 2 GPUs and I run into the same problem when I start 2 jobs each using one GPU (kinda pointless to use DDP in this case but it's easy for prototyping and fixing problems such as this). We also have a compute cluster with nodes that have 4 GPUs. Imagine two people submitting jobs that run DDP on 2 GPUs each. Then one of the jobs will crash because the other has already initialized DDP on that node (I tested it today for jobs of mine).
I am not at work right now, I will try some things and let you know. If you know how to solve the 'what ip address and port should I set' question that would help me a great deal.
Best,
Fabian

**Comment by FabianIsensee** [comment]
Hah

`python -m torch.distributed.launch --nproc_per_node=1 --master_port=2345 train1.py`

`python1 -m torch.distributed.launch --nproc_per_node=1 --master_port=1234 train2.py`

this works!
(python1 is an alias for` CUDA_VISIBLE_DEVICES=1 python`)

Now I just have to test this for our servers and also figure out a way to set a unique port for all jobs, but !thanks! for you help =)

**Comment by mcarilli** [comment]
Oh nice!  I think the master IP address and port are just to help the various processes rendezvous.  In other words, on a single node, I THINK you can specify any IP and port that aren't currently in use and it will work.  Not 100% sure about that though.

**Comment by mcarilli** [comment]
Please keep me posted on whether this works for your DGX2s as well.

**Comment by FabianIsensee** [comment]
I will test this out tomorrow and let you know =)

**Comment by mcarilli** [comment]
https://github.com/NVIDIA/apex/issues/162#issuecomment-466144252  You're only setting CUDA_VISIBLE_DEVICES explicitly for the second job, and allowing the first job to default to gpu0.  For safety, it's probably worthwhile to set CUDA_VISIBLE_DEVICES explicitly for both jobs.

**Comment by FabianIsensee** [comment]
Hi,
yes i'll set CUDA_VISIBLE_DEVICES for all processes. That was just a quick test =) I usually always do that.

Good news: I can confirm that this works for the dgx2, at least when you set --master_port for each training. Changing  --master_addr does NOT work! As soon as you set something else than the default you'll get an error message.

I encountered 2 problems that might be of interest to you:
1) unrelated, but potentially important: Current torch-nightly (from today) is incompatible with amp:
```
Traceback (most recent call last):
  File "DDP.py", line 24, in <module>
    init_method='env://')
  File "/home/isensee/dl_venv/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 354, in init_process_group
    store, rank, world_size = next(rendezvous(url))
  File "/home/isensee/dl_venv/lib/python3.6/site-packages/torch/distributed/rendezvous.py", line 143, in _env_rendezvous_handler
    store = TCPStore(master_addr, master_port, world_size, start_daemon)
ValueError: host not found: Name or service not known
Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/isensee/dl_venv/lib/python3.6/site-packages/torch/distributed/launch.py", line 238, in <module>
    main()
  File "/home/isensee/dl_venv/lib/python3.6/site-packages/torch/distributed/launch.py", line 234, in main
    cmd=process.args)
subprocess.CalledProcessError: Command '['/home/isensee/dl_venv/bin/python', '-u', 'DDP.py', '--local_rank=0']' returned non-zero exit status 1.

```
rollback to stable fixed this problem.

2) Trainings with DDP that crash sometimes leave behind zombie processes. That is super annoying as I either have to manually crawl htop to kill them OR use `killall python` (which is not a good idea if there are other trainings running)

I also encountered another issue that I have with DDP. This is off-topic and not related to apex, but I'll just post it anyways in the hope that you know a quick solution to this. If you don't have the time, don't worry and just ignore this:
From my understanding, each worker in DDP sees only it's own inputs and outputs to the model. The losses are computed on each worker individually, the backprops are done on each worker individually and then the gradients are finally communicated and averaged between the workers. This is fine if you are using a loss function such as a CE in image classification where the loss of each sample in the minibatch is independent and you can just either average the loss or the gradients. In medical image segmentation (which is my specialty) this is problematic. The loss function we typically use is the dice loss (similar to IoU loss). Due to the size of our images and memory constraints, we usually train on patches. One nice trick we do to stabilize training is to compute the dice loss not for each sample indepentently, but using the patches in our minibatch as a larger pseudo-volume by accumulating true positive, false positive and false negative terms (all in some continuous, non discrete form) over the batch. Here is an example to illustrate that:
```
output = net(input) # output has shape (3, 2, 128,128,128), 3 samples in the batch, 3 classes, 128³ patch size)
tp, fp, fn = get_tp_fp_fn(input, reference) # tp, fp and fn each have shape (3, 2) (accumulated over spatial dimension)
# you can also make tp, fp, fn computation part of the forward pass to reduce GPU-GPU communication and reduce load on GPU0
# now accumulate tp/fp/fn over examples
tp = tp.sum(0)
fp = fp.sum(0)
fn = fn.sum(0)
loss = 2 * tp / (2 * tp + fp + fn)
```
In nn.DataParallel this was not a problem at all because only the execution of `output = net(input)` was parallelized over GPUs. i.e. output still had shape (3, 2, 128,128,128) (each GPU outputs (1, 2, 128, 128, 128) and the outputs are stacked together). Now with DDP I would start 3 workers, each with a batch size of 1. The outputs are never stacked and I do not have access to the tp, fp, and fn values of the other workers.
All I need to have to get this running is some function that in the forward pass does a all_gather on these terms and in the backward pass redistributes the gradients to the workers. I looked at your implementation of SyncBatchnormFunction got get some idea of how this needs to be implemented but as I have never written my own pytorch Function I could not get it to work so far :-)

Best,
Fabian

**Comment by mcarilli** [comment]
It doesn't look to me like either 1. or 2. are related to Apex or Amp.  

2. is a known issue with `torch.distributed.launch`.  We've wanted to push a fix for this to Pytorch upstream for a month or so but I haven't found the cycles yet.  At the moment, I typically kill zombie processes manually with this `ps aux | grep python | awk '{ print $2 }' | xargs kill -9`...not ideal but this is on our radar to fix.

As for 1., it also looks like some error with Pytorch's distributed initialization.  I doubt it's related to Amp, because amp is just mixed precision casts and things that occur purely within each process.  It might be related to Apex DDP, but I don't see anything Apex DDP-related.  The stack trace comes from the `torch.distributed.init_process_group` call, so you haven't even gotten to Apex DDP yet.  If `torch.distributed.init_process_group` is failing on the current Pytorch nightlies, this is definitely worth filing an upstream issue, and checking if there are any existing issues.

**Comment by mcarilli** [comment]
As for gathering the inputs, that sounds like hte right idea:  a custom autograd function with an allgather in forward and a scatter in backward.  I've never written a Function that does this specifically but the [documentation](https://pytorch.org/docs/master/notes/extending.html) is really good.  Your forward pass would take one input (this process' output) then do the allgather and return the allgathered thing, then each process computes the loss on the allgathered thing.  So for the post-allgather forward loss computation, the separate processes would be duplicating effort, but that's probably negligible relative to the overhead of the training as a whole.  Your backward pass would take an incoming gradient that would, I think, be bigger than is necessary for just that process, then return a slice of that incoming gradient based on torch.distributed.get_rank.  I don't think you'd need an additional comm in backward, just have each process take and pass along the right slice.  I THINK this mode of operation will end up giving you gradients that are bitwise identical to the control flow that DataParallel would enact (the difference being that DataParallel would only compute the loss itself on gpu 0, then scatter it, while in DistributedDataParallel, the computation of hte loss on the allgathered output is duplicated on each process, but in the backward pass, each process chooses to take just a slice of the resulting gradients, the same slice that it would have been given by DataParallel's backward scatter).  I've never done this before myself (try to duplicate DP-like control flow with DDP) so i'm not 100% sure of all this...let me know if it sounds sensible.  Specifically, what I imagine is:

```
# Within each DDP process:
output = net(input) # output currently has shape (1, 2, 128,128,128)
output = awesome_allgather_function(output) # custom autograd function that does an allgather in forward, and picks a slice of the grads (grad_output[torch.distributed.get_rank()]) to send along in backward
# output is now (3, 2, 128,128,128) in EACH PROCESS
# Each process duplicates the same loss computation, but that's ok,
# because they'll pick out the right slice of gradients in backward
tp, fp, fn = get_tp_fp_fn(output, reference) # tp, fp and fn each have shape (3, 2) (accumulated over spatial dimension)
# you can also make tp, fp, fn computation part of the forward pass to reduce GPU-GPU communication and reduce load on GPU0
# now accumulate tp/fp/fn over examples
tp = tp.sum(0)
fp = fp.sum(0)
fn = fn.sum(0)
loss = - 2 * tp / (2 * tp + fp + fn) # loss has shape (num_classes,)
loss = loss.mean()
```

**Comment by mcarilli** [comment]
If `tp, fp, fn` are created independently from each process's `output`, instead of applying the allgather function to the `output`, maybe it makes more sense to insert an allgather function applied to each of the `tp, fp, fn` after each process creates its `tp, fp, fn`, to reduce the data that must be communicated.   Test what gives best performance.  Collectives are expensive so one collective with a lot of data MAY be cheaper than 3 collectives with a small amount of data.  Of course, you also rig some additional logic to cat the `tp, fp, fn` comms into one collective.

For mimicking DataParallel on a single tensor, I think your custom function could be as simple as 
```
class awesome_allgather_function(torch.autograd.Function):
    @staticmethod
    def forward(ctx, input):
        # create a destination list for the allgather.  I'm assuming you're gathering from 3 workers.
        allgather_list = [torch.empty_like(input) for _ in range(3)] 
        torch.distributed.all_gather(input, allgather_list)
        return torch.cat(allgather_list, dim=0)

    @staticmethod
    def backward(ctx, grad_output):
        # We'll receive gradients for the entire catted list, so to mimic DataParallel,
        # return only the slice that corresponds to this process's input:
        return grad_output[torch.distributed.get_rank()]
```
Not 100% sure this is exactly what you want, but it should point you in the right direction.  I think you can apply this to either the `output` or the `tp, fp, fn`.

**Comment by FabianIsensee** [comment]
HI there,
sorry for not responding earlier. I very much appreciate you effort info my all_gather problem! The last solution you posted sounds exactly what I need! I filled up all available GPUs on Friday, so no possibility to test this right now but I will as soon as I can and then report back if it works =)
Like you already pointed out, I also believe it will be a lot faster to collect tp, fp and fn instead of output. Output is (1,3,224,224,224) per GPU, so about 129MB in fp32. By all_gathering output first, then computing tp fp and fn on each GPU, you will increase the memory usage per GPU by a lot (think 16 GPUs -> 129MB * 16 = 2GB just for the all_gather) and you need to do computations on huge tensors on each worker to get tp fp and fn. Maybe all_gather output is faster for 2 or 4 GPUs, but it certainly won't scale as well as all_gathering tp, fp and fn.
Once again, thank you so much!

**Comment by FabianIsensee** [comment]
```
class awesome_allgather_function(autograd.Function):
    @staticmethod
    def forward(ctx, input):
        world_size = distributed.get_world_size()
        # create a destination list for the allgather.  I'm assuming you're gathering from 3 workers.
        allgather_list = [torch.empty_like(input) for _ in range(world_size)]
        distributed.all_gather(allgather_list, input)
        return torch.cat(allgather_list, dim=0)

    @staticmethod
    def backward(ctx, grad_output):
        grads_per_rank = grad_output.shape[0] // distributed.get_world_size()
        rank = distributed.get_rank()
        # We'll receive gradients for the entire catted forward output, so to mimic DataParallel,
        # return only the slice that corresponds to this process's input:
        return grad_output[(rank * grads_per_rank) : ((rank + 1) * grads_per_rank)]
```
I modified it a little to allow for different number of `--nproc_per_node` and different shapes of tp, fp, fn but all in all your solution was already pretty much all I needed =)
Once again, big thanks! I am learning so much these days!!

**Comment by mcarilli** [comment]
Great, thanks for keeping me updated.  This DataParallel-mimicking strategy may end up being useful for other people in the future.  

Is performance still significantly better than DataParallel?  (I'm not sure if you're also coalescing the the `tp, fp, fn` comms into a single all_gather but that may make it faster if you think comms are a bottleneck).

**Comment by FabianIsensee** [comment]
Honestly, I could not identify a big performance gap between DP and DDP yet. Here is what my scaling looks like. The time here means 'time for 16 iterations' and is measured after running 8 warm-up iterations.
Time is in seconds. batch size per GPU was 2 with a patch size of 64x190x160. Architecture is some 3D Unet network (plain 3d convs, no residual or dense connections)

> num gpus | DDP | DP | DDP (cudnn.benchmark=True) | DP  (cudnn.benchmark=True)
> -- | -- | -- | -- | --
> 1 | 28.08 | 28.05 | 20.36 | 20.28
> 2 | 28.09 | 29.03 | 20.46 | 20.76
> 4 | 28.96 | 30.93 | 20.72 | 21.38
> 8 | 28.97 | 31.51 | 21.39 | 21.64
> 16 | 28.57 | 30.48 | 20.81 | 22.80

An ideal result would be the same run time regardless of GPUs as I increase the total batch size with each GPU (so 1 GPU = batch size 2, 4 GPUs = batch size 8 etc)

My impression is that because my patch size is so large and the GPU time per iteration is so long (almost 2s) that the communication or GIL overhead plays only a minor role. Overall the performance with DDP is better than with DP, but not by much

**Comment by FabianIsensee** [comment]
> Is performance still significantly better than DataParallel? (I'm not sure if you're also coalescing the the `tp, fp, fn` comms into a single all_gather but that may make it faster if you think comms are a bottleneck).

I compared two trainings on my local PC at home (just 2 GPUs). One with the awesome_allgather function and one that just skipped it. Both experiments were within margin of error equally fast. The tensors that needed to be communicated were of shape 107x3, so really quite small (I used 2d convs and a batch size of 107 per GPU, I reduced tp, fp and fn to a nominator term and a denominator term so that I have to allgather only 2 instead of 3 tensors)

**Comment by FabianIsensee** [comment]
Can I close this issue or is there anything else you would like to discuss?
Once again, many thanks!
Fabian

**Comment by mcarilli** [comment]
No, you can close it.

**Comment by wentaozhu** [comment]
@FabianIsensee I am using Apex pytorch o1 for 3D UNet with batch normalization and find the dice loss is pretty unstable and I guess the reason is because the fp16 cannot represent big numbers of fp, tp etc. And o0 solves the issue. Did you encounter this issue and how did you solve it? Thank you so much!

**Comment by FabianIsensee** [comment]
Hi,
the Dice loss is stable for me - no complaints there. I am sure there is a way to force certain calculations to be done in float, maybe just .float() all the inputs?
Best,
Fabian

**Comment by FabianIsensee** [comment]
Could you provide some image sizes etc? Maybe your images are simply larger than mine :-)
Also what variant of the Dice loss are you using?

**Comment by wentaozhu** [comment]
The image size is about 256 x 256 x 256. I use soft dice loss  in https://www.jeremyjordan.me/semantic-segmentation/  What is your successful story?

**Comment by FabianIsensee** [comment]
Interesting that Dice loss is basically my old lasagne/theano implementation ported to pytorch :-) 
I don't see why it should not be working. If your gradients are overflowing amp will give you lots of warnings. If fp16 cannot represent numbers that are large enough you will probably end up with nans in the loss. Anything that does not give warnings or nans should be working. Have you tried forcing the input of your loss to .float()? (just set y_pred = y_pred.float() and y_true = y_true.float() at the beginning of the loss function)? Is the loss function part of the network module or is it separate? (I am asking because I think everything that you pass to amp.initialize will be run in mixed precision. If the loss function is separate then it should run in fp32 regardless. Maybe @mcarilli has some insights? I am just speculating.)
It also always helps to provide a code example that underlines the problem

**Comment by FabianIsensee** [comment]
Also your loss does the squaring in the denominator which I do not fully agree with. I am not using that. Try replacing this

    denominator = np.sum(np.square(y_pred) + np.square(y_true), axes)

with this

    denominator = np.sum(y_pred + y_true, axes)

(just noticed this is a numpy implementation, you need to adapt this accordingly depending on what framework you are using)

**Comment by wentaozhu** [comment]
Thank you so much! I will try it again.

**Comment by alexfoo-dw** [comment]
Hi @FabianIsensee , thanks for your input on this. I have tried launching several independent jobs with separate invocations (A,B,C,D…) on different GPUs and they work fine.

Now in order to utilize my GPU ram better, I have tried overloading several experiments on the same GPUs (e.g. A, B with DDP on GPUs 0,1) but notice a significant slow-down during backprop and slight slow-down for forward pass for both A and B despite getting expected training results. The total spawned processes are also well within the number of CPU cores. Can I check if there is any way to optimize this for speed or is this a bottleneck on how NCCL passes messages between GPUs?

Thanks in advance!
